### DISCUSSION TITLE
Improvements

### DISCUSSION THREAD

**Discussion by Akascape** [PaulleDemon/tkVideoPlayer#26] [Ideas]
Hello @PaulleDemon,
I just found the correct way to seek any particular frame using pyav:

```python
frame_num = 1000 # the frame I want
framerate = container.streams.video[0].average_rate # get the frame rate
time_base = container.streams.video[0].time_base # get the time base
sec = int(frame_num/framerate) # timestamp for that frame_num
container.seek(sec*1000000, whence='time', backward=True)  # seek to that nearest timestamp
frame = next(container.decode(video=0)) # get the next available frame
sec_frame = int(frame.pts * time_base * framerate) # get the proper key frame number of that timestamp

for _ in range(sec_frame, frame_num):
    frame = next(container.decode(video=0))
return frame
```

I properly implemented a new `seek_frame` method in tkvideoplayer, then I tested it:
Here is demo example:

https://user-images.githubusercontent.com/89206401/228515652-abe137d0-6823-4c56-ba5c-8eb8292e0182.mp4

There is no memory leaks or lags while seeking, only a little issue with cpu usage if we continuously keep on seeking frames which can be fixed by adding a time delay.

Basically the modification I did:
- added seek_frame method
- added any_frame parameter in seek method (useful for some video file)
- added new event: << FrameChanged >>
- fixed a loading issue when pyav crashes
- added pause option for seek method, in case if someone wants to seek while pause.
- other small improvements

So, should I give the pull request so that you can review it?

**Comment by PaulleDemon** [comment]
@Akascape sure you can, but I might not be able to review it until next month as I am quite packed these days.

**Reply by Akascape** [reply]
@PaulleDemon 
>sure you can, but I might not be able to review it until next month as I am quite packed these days.

No problem, I submitted the PR: https://github.com/PaulleDemon/tkVideoPlayer/pull/27
